### PR TITLE
SI-8047 change fresh name encoding in quasiquotes to avoid symbol owner corruption

### PR DESCRIPTION
**Pull request by densh**
Previously a following encoding was used to represent fresh names that should be created at runtime of the quasiquote:

```
build.withFreshTermName(prefix1) { name$1 =>
  ...
    build.withFreshTermName(prefixN) { name$N =>
      tree
    }
  ...
}
```

It turned out that this encoding causes symbol corruption when tree defines symbols of its own. After being spliced into anonymous functions, the owner chain of those symbols will become corrupted.

Now a simpler and probably better performing alternative is used instead:

```
{
  val name$1 = build.freshTermName(prefix1)
  ...
  val name$N = build.freshTermName(prefixN)
  tree
}
```

Here owner stays the same and doesn’t need any adjustment.

### CODE REVIEW

**Comment by densh** [comment]
review @retronym @xeno-by

**Comment by xeno-by** [comment]
LGTM

File: test/files/run/si8047.scala [hunk 1]
```diff
@@ -0,0 +1,4 @@
+object Test extends App {
```
Comment (retronym) [inline test/files/run/si8047.scala]: Test file should be named `t8047.scala`

File: test/files/run/si8047.scala [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+  import scala.reflect.runtime.universe._
+  println(q"do ${typeOf[String]} while(true)")
```
Comment (retronym) [inline test/files/run/si8047.scala]: Is `typeOf` essential here? Or just is splicing something like `() => ...` sufficient to trigger the problem? I still can't quite connect the dots to say precisely where the bug came from; a more focussed test and elaborated commit comment would help.

Comment (xeno-by) [inline test/files/run/si8047.scala]: I think anything symbol-bearing will do.

Comment (densh) [inline test/files/run/si8047.scala]: After fiddling with it a bit I've minimized it down to:

``` scala
object Test {
  q"do ${ val x = 1; x } while(true)"
}
```

Here x's owner is outer Test scope. Now you expand the quasiquote and get:

``` scala
object Test {
  build.withFreshTermName("doWhile")(n => 
    LabelDef(n, List(), 
      Block(
        List({ val x = 1; x }), 
        If(Literal(Constant(true)), Apply(Ident(n), List()), Literal(Constant(())))))
}
```

Here the proper owner is anonymous function, not the Test. Hence symbol corruption. In new encoding this is represented as:

``` scala
object Test {
  { 
    val n = build.freshTermName("doWhile")
    LabelDef(n, List(), 
      Block(
        List({ val x = 1; x }), 
        If(Literal(Constant(true)), Apply(Ident(n), List()), Literal(Constant(()))))
  }
}
```

Owner stays the same and life is good again.

File: src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala [hunk 3]
```diff
@@ -33,8 +33,16 @@ trait Reifiers { self: Quasiquotes =>
...
-     *    a sequence of nested withFreshTermName/withFreshTypeName calls which are required
-     *    to force regeneration of randomly generated names on every evaluation of quasiquote.
+     *    a block which starts with a sequence of vals that correspond
+     *    to fresh names that has to be created at evaluation of the quasiquote
+     *    and ends with reified tree:
+     *
+     *      {
+     *        val name1: universe.TermName = universe.build.freshTermName(prefix1)
+     *        ...
+     *        val nameN: universe.TermName = universe.build.freshTermName(prefixN)
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala]: Should we use `name$1` instead? Even though you don't `resetLocalAttrs` on the splicee (AFAIK), things might not typecheck anymore if the expansion of the quasiquote was passed through an identity macro that `typecheck(resetLocalAttrs(tree)`.

Comment (densh) [inline src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala]: Names are created as `c.freshName[TermName](nme.QUASIQUOTE_NAME_PREFIX)` so they are both semi-fresh and defined by a prefix that is a configurable constant. I guess I'll update the comment to clarify this.

**Comment by adriaanm** [comment]
I'll let this simmer until we get that second LGTM

**Comment by densh** [comment]
Updated with minimized test case, better comments and proper test file names. /cc @retronym

**Comment by retronym** [comment]
LGTM

**Comment by adriaanm** [comment]
Rebuilding the test suite. Will merge once https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/4325/console succeeds

**Comment by adriaanm** [comment]
No idea why, but the test suite genuinely seems to hang on this PR.

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Comment by xeno-by** [comment]
@adriaanm @retronym The hang is gone. Since it was produced by `jvm/reactor-exceptionOnSend.scala`, I guess, the reasons for it lie outside this PR. We have a double LGTM, so let's merge!

**Pull request merged**